pulsar/impl_message.go

```diff
@@ -126,3 +142,36 @@ func (msg *message) EventTime() time.Time {
 func (msg *message) Key() string {
 	return msg.key
 }
+
+func newAckTracker(size int) *ackTracker {
+	var batchIDs *big.Int
+	if size <= 64 {
+		shift := uint32(64 - size)
+		setBits := ^uint64(0) >> shift
+		batchIDs = new(big.Int).SetUint64(setBits)
+	} else {
+		batchIDs, _ = new(big.Int).SetString(strings.Repeat("1", size), 2)
+	}
+	return &ackTracker{
+		size:     size,
+		batchIDs: batchIDs,
+	}
+}
+
+type ackTracker struct {
+	sync.Mutex
+	size     int
+	batchIDs *big.Int
+}
+
+func (t *ackTracker) ack(batchID int) {
+	t.Lock()
+	defer t.Unlock()
+	t.batchIDs = t.batchIDs.SetBit(t.batchIDs, batchID, 0)
+}
+
+func (t *ackTracker) cleared() bool {
```

Comment:
Maybe we can use a term that's more specific to acks rather than bitset. eg; 
 * `IsComplete()` 
 * `EverythigWasAcked()`

Another option could be to return bool from `ack(batchID int)`. That will also avoid getting the mutex twice each time.

Reply:
I'll add a bool to the return value to ack(batchID int) bool and change the cleared() -> completed(). Thoughts?